File: Data Structures & Algorithms/buy-and-sell-crypto/submission-1.java
class Solution {
    public int maxProfit(int[] prices) {
        // assume we buy the first price
        int profit = 0, buy = prices[0];

        for (int i=1;i<prices.length;i++){
            // if current price is higher then buying price -> profit
            if (prices[i] > buy){
                profit = Math.max(profit, prices[i] - buy);
            }
            // found lower price then current assumed buying price -> update assume buying price to the lower price
            else {
                buy = prices[i];
            }
        }
        return profit;
    }
}

File: Data Structures & Algorithms/best-time-to-buy-and-sell-stock-ii/submission-1.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        cache = {}
        
        def dfs(idx, bought):
            if idx == len(prices):
                return 0
            if (idx, bought) in cache:
                return cache[(idx, bought)]

            do_nothing = dfs(idx + 1, bought)
            if bought:
                res = max(do_nothing, prices[idx] + dfs(idx + 1, False))
            else:
                res = max(do_nothing, -prices[idx] + dfs(idx + 1, True))
            cache[(idx, bought)] = res
            return res

        return dfs(0, False)